Makefile

```
@@ -8,59 +8,102 @@ MAIN_SRC = main.cob
 OBJECTS_DIR = out
 OBJECTS = $(patsubst src/%.cob, out/%.o, $(SRC))
 CPY_DIR = src/copybooks
-CPY = $(wildcard src/copybooks/*.cpy)
 CPP_SRC = $(wildcard cpp/*.cpp)
 CPP_HEADERS = $(wildcard cpp/*.h)
 CPP_OBJECTS = $(patsubst cpp/%.cpp, out/cpp/%.o, $(CPP_SRC))
 
 BIN = cobolcraft
 
+all: $(BIN) data
+
 # Data extraction from Mojang's server.jar
 SERVER_URL = https://piston-data.mojang.com/v1/objects/4707d00eb834b446575d89a61a11b5d548d8c001/server.jar
 SERVER_JAR_EXTRACTED = data/versions/1.21.4/server-1.21.4.jar
 
 # Test sources and binary
-TEST_SRC = test.cob $(wildcard tests/*.cob)
+TEST_SRC = $(wildcard tests/*.cob)
+TEST_OBJECTS = $(patsubst tests/%.cob, out/tests/%.o, $(TEST_SRC))
 TEST_BIN = test
 
-# GnuCOBOL >=3.2 added functions that drastically improve performance. Set this to 32 to enable them.
-GCVERSION=31
-
-# Common compiler options
-COBC_OPTS = -O2 -debug -Wall -fnotrunc --free -fstatic-call -DGCVERSION=$(GCVERSION)
-
-.PHONY: all clean data run test
-
-all: $(BIN) data
-
-clean:
+# GnuCOBOL >=3.2 added functions that drastically improve performance, query version (may one sub-digit) automatically
+GCVERSION := $(shell $(COBC) --version | head -n1 | sed 's/.*\s\([0-9]\)\.\([0-9]\).*/\1\2/g' )
+
+# Common compiler options, note: COB_FLAGS can be user-specified
+COBC_OPTS = --free -DGCVERSION=$(GCVERSION) -I $(CPY_DIR) $(COB_FLAGS)
+ifneq ($(DEPONLY),Y)
+COBC_OPTS += -g -O2 --debug -Wall -fnotrunc -fstatic-call
+else
+COBC_OPTS += --fsyntax-only -w
+endif
+
+ifeq ($(shell test $(GCVERSION) -ge 32; echo $$?),0)
+COB_MTFLAGS = -MF $(basename $@).d -MT $@
+GLOBALDEPS =
+ifneq ($(DEPONLY),Y)
+# include dependency files for all existing object files
+-include $(OBJS:.o=.d)
+else
+.PHONY: $(OBJECTS) $(TEST_OBJECTS)
+endif
+else
+# if dependency generation is not available: use global copy dependency
+COB_MTFLAGS = 
+GLOBALDEPS = $(wildcard src/copybooks/*.cpy)
+endif
+
+.SUFFIXES:
+.PHONY: all clean cobclean data run test
+
+cobclean:
 	rm -rf $(OBJECTS_DIR)
 	rm -f $(BIN)
-	rm -f $(TEST_BIN)
+	rm -f $(TEST_BIN) $(TEST_OBJECTS)
```

```suggestion
	rm -f $(TEST_BIN)
```

(test objects will already be removed due to `OBJECTS_DIR` being removed)